lang/cs/Org.Apache.REEF.Common/Telemetry/DriverMetrics.cs

```diff
@@ -15,25 +15,47 @@
 // specific language governing permissions and limitations
 // under the License.
 
-using System;
+using StringMetric = Org.Apache.REEF.Common.Telemetry.MetricClass<string>;
 
 namespace Org.Apache.REEF.Common.Telemetry
 {
     /// <summary>
-    /// A simple driver metrics.
-    /// It contains system state for now.
-    /// It can be extended later to include more driver metrics data.
+    /// Driver metrics implementation that contains the system state.
     /// </summary>
     public sealed class DriverMetrics : IDriverMetrics
     {
-        public DriverMetrics(string systemState, DateTime timeUpdated)
+        internal MetricsData _metricsData;
+
+        public IMetric SystemState
         {
-            SystemState = systemState;
-            TimeUpdated = timeUpdated;
+            get;
         }
 
-        public string SystemState { get; private set; }
+        private string _stateMetricName = "DriverState";
 
-        public DateTime TimeUpdated { get; private set; }
+        public DriverMetrics(string systemState)
+        {
+            _metricsData = new MetricsData();
+            SystemState = new StringMetric(_stateMetricName, "driver state.", systemState, false);
+            _metricsData.RegisterMetric(SystemState);
+        }
+
+        public IMetrics GetMetricsData()
+        {
+            return _metricsData;
+        }
+
+        public T CreateAndRegisterMetric<T>(string name, string description, bool keepUpdateHistory) 
+            where T : MetricBase, new()
+        {
+            var metric = new T
```

Comment:
Minor nit - You can make this be T(name, description, keepUpdateHistory). Then you can have private setters on MetricBase.

Reply:
So this is a restriction on C# generics that has been discussed here:
https://stackoverflow.com/questions/1682310/c-sharp-generics-problem-newing-up-the-generic-type-with-parameters-in-the-con

If the constructor were to be called directly, it'd need to be passed  as a parameter, like so:
        public T CreateAndRegisterMetric<T>(Func<string, string, bool, T> del, string name, string description, bool keepUpdateHistory) where T : MetricBase
        {
            var me = del(name, description, keepUpdateHistory);
            return me;
        }
which I don't think is a good idea since there's no way to ensure the function that gets passed in is actually the constructor that we want-- really method that takes three arguments and returns T would fly.
---
Another option is 

var metric = Activator.CreateInstance(typeof(T), name, description, keepUpdateHistory) as T

but here we'd trust that the derived class has a constructor that takes three arguments of the correct types. It's safer than the first method because at least we'll know it's a constructor. If no constructors match, this will throw. For all of our primitive types this would work, but if someone were to derive from the MetricClass<T> type, they'd need to add their own constructor, otherwise their metric wouldn't be able to register.
---
All this is to say, internal setters seemed to me like a reasonable compromise to avoid user error when creating and registering metrics. Comments and criticism welcome :P